src/gorilla.c

```diff
@@ -0,0 +1,451 @@
+    } else {
+        CHECKSPACE(chunk, 5 + 64 + 1);
+        appendBits(bins, bit, 0x1f, 5);
+        appendBits(bins, bit, doubleDelta.u, 64);
+    }
+    chunk->prevTimestampDelta = curDelta;
+    chunk->prevTimestamp = timestamp;
+    return CR_OK;
+}
+
+static ChunkResult appendV(CompressedChunk *chunk, double value) {
+    union64bits val;
+    val.d = value;
+    u_int64_t xorWithPrevious = val.u ^ chunk->prevValue.u;
+
+    binary_t *bins = chunk->data;
+    globalbit_t *bit = &chunk->idx;
+    
+    // CHECKSPACE already checked for 1 extra bit availability in appendTS.
+    // Current value is identical to previous value. 1 bit used to encode.
+    if (xorWithPrevious == 0) {
+        appendBits(bins, bit, 0, 1);
+        return CR_OK;   
+    }
+    appendBits(bins, bit, 1, 1);
+
+    u_int64_t leading  = LeadingZeros64(xorWithPrevious);
+    u_int64_t trailing = TrailingZeros64(xorWithPrevious);
+  
+    // Prevent over flow of DOUBLE_LEADING
+    if (leading > 31) { 
+        leading = 31; 
+    }
+
+    localbit_t prevLeading = chunk->prevLeading;
+    localbit_t prevTrailing = chunk->prevTrailing;  
+    assert(leading + trailing <= BINW);
+    localbit_t blockSize = BINW - leading - trailing;
+    u_int32_t expectedSize = DOUBLE_LEADING + DOUBLE_BLOCK_SIZE + blockSize;
+    assert(prevLeading + prevTrailing <= BINW);
+    localbit_t prevBlockInfoSize = BINW - prevLeading - prevTrailing;
+/*
+ * First bit encodes whether previous block parameters can be used since
+ * encoding block-size requires 5 + 6 bits.
+ * 
+ * If previous block size is used and the block itself is being appended.
+ * 
+ * Else, number of leading zeros in inserted followed by trailing zeros.
+ * Then the value is the block is being appended. 
+ */
+    if (leading >= chunk->prevLeading && 
+        trailing >= chunk->prevTrailing &&
+        expectedSize > prevBlockInfoSize) 
+    {
+        CHECKSPACE(chunk, prevBlockInfoSize + 1);
+        appendBits(bins, bit, 0, 1);    
+        appendBits(bins, bit, xorWithPrevious >> prevTrailing, prevBlockInfoSize);
+    } else {
+        CHECKSPACE(chunk, expectedSize + 1);
+        appendBits(bins, bit, 1, 1);
+        appendBits(bins, bit, leading, DOUBLE_LEADING);
+        appendBits(bins, bit, blockSize - DOUBLE_BLOCK_ADJUST, DOUBLE_BLOCK_SIZE);            
+        appendBits(bins, bit, xorWithPrevious >> trailing, blockSize);  
+        chunk->prevLeading = leading;
+        chunk->prevTrailing = trailing;
+    }
+    chunk->prevValue.d = value;
+    return CR_OK;
+}
+
+ChunkResult CChunk_Append(CompressedChunk *chunk, timestamp_t timestamp, double value) {
+    assert(chunk);
+
+    if (chunk->count == 0) {
+        chunk->baseValue.d   = chunk->prevValue.d = value;
+        chunk->baseTimestamp = chunk->prevTimestamp = timestamp;
+        chunk->prevTimestampDelta = 0;
+    } else {
+        if (appendTS(chunk, timestamp) != CR_OK) return CR_END;
+        if (appendV (chunk, value) != CR_OK) return CR_END;
```

Comment:
@ashtul how do you handle the TS being able to write to the chunk but the value doesn't have space?
This might impact the reading algorithem that expects both timestamp and value.

Reply:
By checking if I have another sample in the chunk. Line 445. So if indeed we wrote the timestamp but not the value, those bits are 'dirty' but we will never read them.

I won't impact the reading since every iterator has to start from the beginning